### ISSUE TITLE
Search facets should support untagged (was: should be unchecked initially)

### ISSUE THREAD

**Issue by cillianderoiste** [ploneintranet/ploneintranet#633]
With SOLR enabled:
1. Create 2 Documents: "Doc A" with the tag TagA and "Untagged Doc"
2. Search for "Doc"

The search facet for Tags will show "TagA" with the box checked.

If you uncheck the box, you still get two results. If you check it again you only get the document which has been tagged with TagA.

I guess that all facets should be unchecked initially.

**Comment by cillianderoiste** [comment]
Hi @cornae, @mattss mentioned that the facet checkboxes are checked initially in the prototype and I see there is also a checkbox for "All other tags" which we haven't implemented (yet?). How should this work?

**Comment by cornae** [comment]
The all other tags is a combination checkbox for all tags that are not in the list. So if it's ticked, all tags that are not in the checklist are shown. If unticked, those are hidden, but tags that are in the tag list might still be shown.

**Comment by cillianderoiste** [comment]
@cornae thanks, I guess that means we only include the top 8 tags, and that "all other tags" also includes "no tags", and we'll need this option for any facet which could be empty. So in the example above:
1. Search for "Doc"
2. Both "TagA" and "All other tags" are checked
3. Results are "Doc A" and "Untagged Doc"
4. Click Tags->"None"
5. No results
6. Check "All other tags"
7. Get "Untagged Doc"
8. Check "TagA" and uncheck "All other tags"
9. Get "Doc A"

I see that the tags aren't ordered alphabetically, so I guess they are ordered by popularity in the search results (but probably are ordered alphabetically where they have the same amount of results).

Does that sound right?

It seems a bit odd to me to include untagged items in the results along with less popular tags, so I hope I've understood this correctly.

**Comment by cornae** [comment]
The untagged one is an interesting one. I could also imagine that we add an extra box for 'untagged'. That could solve some issue. @gyst, How do you feel about that?

**Comment by gyst** [comment]
Ok with me.

On August 10, 2015 2:08:02 PM GMT+02:00, "Cornelis G. A. Kolbach" notifications@github.com wrote:

> The untagged one is an interesting one. I could also imagine that we
> add an extra box for 'untagged'. That could solve some issue. @gyst,
> How do you feel about that?
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/ploneintranet/ploneintranet/issues/633#issuecomment-129423827

## 

Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by cillianderoiste** [comment]
@mattss I think this one will need your attention.

**Comment by gyst** [comment]
I'm also fine with supporting untagged documents as part of the 'all other tags' option, which does need to be implemented.

**Comment by gyst** [comment]
@mattss This is a bug in search that I'd like to get fixed before we release Venus. Any chance of that happening this week?
- We need an "Other (or none)" checkbox on all facets that have dangling items
- The [None] button on a facet should deselect all checkboxes and result in 0 matches in the search result

**Comment by mattss** [comment]
Have hit a slight brick wall here in that SOLR doesn't appear to support negative queries combined with other queries i.e.
- Show me all untagged documents <- this is fine
- Show me all untagged documents, along with documents tagged with 'banana' <- this is not supported by SOLR (See http://wiki.apache.org/solr/SolrQuerySyntax and the note about 'Pure negative queries (all clauses prohibited).

Talking this through with others here, the consensus is that asking for things with _no_ tags as well as a specific tag makes no sense, and only exists here because of the way the form is laid out. 

Looking again at big sites such as amazon - they default to having all checkboxes _unchecked_ to start with, and checking any box narrows the search. Unfortunately I can't quite foresee the complete UI flow that would result from us reversing this design decision.

Any thoughts or pointers let me know ;)

**Comment by mattss** [comment]
Or we could switch to radio buttons and this would all go away ;)

**Comment by gyst** [comment]
OK clear. The design is inconsistent, since "all" and "none" will return the same result set (no filter on this facet) while the UI suggests a big difference. The checkboxes suggest negative queries ("if I remove tag 'foo' all documents tagged 'foo' will not be shown", whereas OTOH the intent is clearly that a facet searches for OR-ed positive values (all documents tagged 'foo' or 'bar').

I'll need to talk with @cornae about this. Cornelis see the original bug report on top, that's the clearest description of a problematic scenario.

**Comment by cornae** [comment]
I understand the thread and the problem, except I don't see how the design would be inconsistent. The intent of the design is that hitting none would design-wise result in no results. I see no inconstancies in the there. 

The other major design paradigm is the one where selecting nothing is the same as no filters applied and therefor all is shown. Selecting one filter means exclude everything that doesn't meet the selected criterium. That paradigm is actually more  advanced search, but some people call it filtering when it gives a live result. 

A third popular paradigm is the narrowing down method, particularly popular with sites that sell things, such as Amazon. Or Bijenkorf.nl, a Dutch web store which is struggling for years with this problem and they've now just starting to try the third solution after having tried the other two. Still the mechanism isn't understood by their major target group so the designers told me ;). Filters and advanced search are still by definition for advanced users. 

I would agree with "asking for things with no tags as well as a specific tag makes no sense". At the same time one could ask oneself if this situation would reasonably occur. In other words, is that specific case a realistic problem? 

Each of the afore mentioned paradigms is a reasonable option, as long as you never mix them. We could go for the none ticked means all type. It will feel natural to some users but unnatural to others. Especially to people who are not used to think in queries. Some people might feel a strong parallel to the tick boxes of their Outlook calendars that behave like our current solution: what is ticked, is shown. 

![Outlook](https://www.officecalendar.com/images/tour/outlook_calendar_share_3.gif)

To avoid this confusion a lot of webshop don't show tick boxes, but for instance accentuate the text or show the selected criteria in a separate box. I would be happy to design such an alternative solution. Of course it will take time from my side and also again from the dev side. Bear in mind though that also there we might find corner cases for which the chosen solution is not optimal. 

The pun of @mattss is actually not a bad solution either :). We use it actually a star and I've used it also in other places like  video search engine I created and a webshop. First radio button 'all' (which is for documents with any or nog tags) and each next radio button is another tag. You limit to x tags max and you don't need boxes for documents with no tags or 'other' tags. It's a typical example where you simplify the 80% use case by sacrificing the 20% use case. 

@gyst Perhaps a call tomorrow to look at our different options?

**Comment by gyst** [comment]
On 22/09/15 23:10, Cornelis G. A. Kolbach wrote:

> The other major design paradigm is the one where selecting nothing is
> the same as no filters applied and therefor all is shown. Selecting one
> filter means exclude everything that doesn't meet the selected
> criterium. That paradigm is actually more advanced search, but some
> people call it filtering when it gives a live result.

I think that is the best and easiest option, as already suggested by
Cillian. Default to showing all boxes unchecked, so that instead of
showing aspects of the current query, they indicate that they're extra
narrowing filters you can apply.

> @gyst https://github.com/gyst Perhaps a call tomorrow to look at our
> different options?

I'm open to that but don't think it's really necessary if you agree with
the above: default to all unchecked. Thanks for your explanations about
the various options, it's clear.

Main remaining question for me is, if we should not remove the buttons [
all ] [ none ] if we're going for the 'all unchecked by default' solution.

## :*CU#

```
Guido Stevens  |  +31.43.3618933  |  http://cosent.nl

s o c i a l   k n o w l e d g e   t e c h n o l o g y
```

**Comment by cornae** [comment]
Moving to the other paradigm means a redesign and part of that will for sure be the removal of the all/none buttons. Those belong only to the flavour we currently use.

**Comment by gyst** [comment]
@mattss I propose that for Venus now we:
- show all checkboxes unchecked by default
- remove the [all] [none] buttons

**Comment by mattss** [comment]
@gyst sounds good - I'll work on this next week. Should be simple to do, just need to rewrite the robot tests!

**Comment by mattss** [comment]
@gyst implemented by #802

**Comment by gyst** [comment]
Yes I saw but ran into issues testing the PR. Am redoing that now with a fresh head :-)
